### PR TITLE
Added support for session logging under dev_1_1 branch this time.

### PR DESCRIPTION
**Pull request by halfmetaljacket**
Otherwise the same as the previous PR. Uses the Python logging module. Ref #279

### CODE REVIEW

File: netmiko/base_connection.py [hunk 1]
```diff
@@ -147,6 +154,7 @@ def _read_channel_expect(self, pattern='', re_flags=0):
             if re.search(pattern, output, flags=re_flags):
                 if debug:
                     print("Pattern found: {} {}".format(pattern, output))
+                self.write_to_session_log(output)
```
Comment (halfmetaljacket) [inline netmiko/base_connection.py]: This line can be removed if line#149 referenced self.read_channel(). I'm not sure if there was a good reason for that though so I didn't touch it.

**Comment by ktbyers** [comment]
@halfmetaljacket I think this is a good idea...just haven't had time to review / merge it in.

**Comment by ktbyers** [comment]
@halfmetaljacket What is the issue that CustomFileHandler is solving?

**Comment by halfmetaljacket** [comment]
The default file handler unconditionally prints a newline at the end of
every log message. This distorts the device output as any required newlines
are already included. Although py3.4+ modified the logging API to
optionally not print newline, there is no such ability currently for
py2.6/2.7. The only backwards-compatible way to prevent printing the
newline is to sub-class the default `FileHandler` class and override the
relevant method.

**Comment by ktbyers** [comment]
@halfmetaljacket Haven't forgotten about this. I am out traveling this week so it might be hard for me to review. I will try to review it next week.

**Comment by ktbyers** [comment]
@halfmetaljacket I think I might want to do it a different way than you did...so I want to look into this some more.

**Comment by halfmetaljacket** [comment]
Ok. If there's any way I can assist, let me know. This was really just my
first go at this, and I wasn't too sure about the best way to present the
API for it.

**Comment by ktbyers** [comment]
@halfmetaljacket It is all good...mostly haven't had time to work on it :-(

**Comment by ktbyers** [comment]
@halfmetaljacket Okay, I did a first past at logging in the four commits on Dec 7 (to develop branch). Here is one of them.

https://github.com/ktbyers/netmiko/commit/a1c1c7a6c62aa3af937d591e0285e0b2c23e3a3a

Right now I kept it pretty small. I need to look some more at what you did.

**Comment by halfmetaljacket** [comment]
Looked through the commits... I like the more library-centric configuration, but that does come with two tradeoffs:
 Not setting up a handler to output to means that it is up to the users of your library to learn logging and how to roll their own handler -not very friendly-, but that does mean it is more flexible for those who wish to figure out what to do. 
Adding in the additional information to each log entry (i.e. your fourth commit) makes debugging much easier, but it kills the "CLI session-log" look that you would get from CLI tools like putty/securecrt/etc.

I like this setup for creating a library debugging log (and might be a very desirable feature just for that), but for the specific use-case of a more traditional "session log" I'm not sure I would use this. Thoughts?

**Comment by halfmetaljacket** [comment]
BTW, if you're still interested in my original take on this, I can modify it to point it to the right branch and make a few minor improvements.

**Comment by halfmetaljacket** [comment]
I have an idea of how to implement both a debug log and an optional session log using a similar setup to what you put together, if you're interested.

**Comment by ktbyers** [comment]
@halfmetaljacket It is pretty easy for a Netmiko consumer to configure logging, they just have to do this:

```
import logging
logging.basicConfig(filename='test.log', level=logging.DEBUG)
```

Is there something I am missing here?

So the SecureCRT/PuTTY log format just makes it somewhat look like you had copied and pasted a CLI session and put it into your logfile--is that correct?

```
I like this setup for creating a library debugging log (and might be a very desirable feature just 
for that), but for the specific use-case of a more traditional "session log" I'm not sure I would 
use this. Thoughts?
```

For debugging code, it is very helpful to me to see the separate read versus write actions (as opposed to having them co-mingled). 

I am somewhat inclined to think it is better to not log like PuTTY/SecureCRT i.e. to log the separate read and writes and make this distinction clear.


What were you thinking as far as the below?

```
I have an idea of how to implement both a debug log and an optional session log using a 
similar setup to what you put together, if you're interested.
```

**Comment by halfmetaljacket** [comment]
My particular use-case is to use session logging as an audit log for my netmiko scripts - a clean output of everything my scripts do to devices that they touch. For this use-case, CLI-like-logs is what I am looking for. If you forgo that scenario to make this more of a debugging log instead, then yeah configuring logging is easy for your users as you mentioned, and I agree that how you have it configured here should work well for that. You could even expand this to replace all of your "if debug" statements with debug logging, avoiding users live editing netmiko code on their systems.

My thinking for that last part would be to implement both a debug log and a separate (or sub-) logger that is used for CLI-like session logging. If you're interested in supporting two separate use cases, the session/audit logging can be set up very similarly to how you have it set up now.

**Comment by ktbyers** [comment]
@halfmetaljacket Okay that makes sense i.e. we had two different use cases.

I definitely want to keep the debug logging. 

How are you thinking of implementing the session-log?


Kirk

**Comment by ktbyers** [comment]
I think I am just going to leave this as is.

If you want otherwise, let me know and we can discuss...I am not really convinced we need the other way.

**Pull request closed without merging**